[Quant] Add ConvTranspose reference module

terrychenism opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #72473

Summary:
Add ConvTranspose reference module

Test Plan:
python3 test/test_quantization.py TestQuantizeEagerOps.test_conv_transpose_op

Reviewers:

Subscribers:

Tasks:

Tags:

terrychenism committed:
[Qunat] Add ConvTranspose reference module

Summary:
Add ConvTranspose reference module

Test Plan:
python3 test/test_quantization.py TestQuantizeEagerOps.test_conv_transpose_op

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

terrychenism commented:
not ready for merge

terrychenism committed:
Update on "[Qunat] Add ConvTranspose reference module"


Summary:
Add ConvTranspose reference module

Test Plan:
python3 test/test_quantization.py TestQuantizeEagerOps.test_conv_transpose_op

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

terrychenism committed:
Update on "[Qunat] Add ConvTranspose reference module"


Summary:
Add ConvTranspose reference module

Test Plan:
python3 test/test_quantization.py TestQuantizeEagerOps.test_conv_transpose_op

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

terrychenism committed:
Update on "[Qunat] Add ConvTranspose reference module"


Summary:
Add ConvTranspose reference module

Test Plan:
python3 test/test_quantization.py TestQuantizeEagerOps.test_conv_transpose_op

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

terrychenism committed:
Update on "[Qunat] Add ConvTranspose reference module"


Summary:
Add ConvTranspose reference module

Test Plan:
python3 test/test_quantization.py TestQuantizeEagerOps.test_conv_transpose_op

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo commented on torch/nn/quantized/_reference/modules/conv.py:
```diff
@@ -217,3 +217,241 @@ def _get_name(self):
+        this is useful when user want to use this module in other backends like Glow.
+    """
+    __annotations__ = {"bias": Optional[torch.Tensor]}
+    _IS_REFERNECE = True
```
nit: `IS_REFERENCE` (fix typo)?

vkuzo commented on test/quantization/eager/test_quantize_eager_ptq.py:
```diff
@@ -74,6 +75,65 @@
+        class M(torch.nn.Module):
+            def __init__(self):
+                super().__init__()
+                self.conv = nn.ConvTranspose2d(1, 1, 1)
```
can we test the 1d and 3d variants as well

terrychenism committed:
Update on "[Qunat] Add ConvTranspose reference module"


Summary:
Add ConvTranspose reference module

Test Plan:
python3 test/test_quantization.py TestQuantizeEagerOps.test_conv_transpose_op

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on torch/ao/quantization/quantize.py:
```diff
@@ -565,7 +565,14 @@ def swap_module(mod, mapping, custom_module_class_mapping):
+            qmod = mapping[type(mod)]
+            if hasattr(qmod, '_IS_REFERENCE') and qmod._IS_REFERENCE:
+                assert mod.qconfig is not None
+                weight_post_process = mod.qconfig.weight()
```
we need to run the weight_process_process with weight of the module:
`weight_post_process(mod.weight)`

jerryzh168 commented on test/quantization/eager/test_quantize_eager_ptq.py:
```diff
@@ -74,6 +75,73 @@
+                x = self.quant(x)
+                x = self.dequant(x)
+                x = self.conv(x)
+                x = self.quant(x)
```
is this is the module for reference conv transpose, we need to use two different QuantStubs since they are recording the scale/zero_point at a specific point

jerryzh168 commented on test/quantization/eager/test_quantize_eager_ptq.py:
```diff
@@ -74,6 +75,73 @@
+                x = self.dequant(x)
+                return x
+
+        class RefM(torch.nn.Module):
```
not sure which one you want to use quantized conv transpose (instead of reference), but the one that is using normal path will have one quant and one dequant:

```
x = self.quant(x)
x = self.convtranspose(x)
x = self.dequant(x)
```
since the quantized convtranspose takes quantized Tensor as input and outputs a quantized Tensor

terrychenism committed:
Update on "[Qunat] Add ConvTranspose reference module"


Summary:
Add ConvTranspose reference module

Test Plan:
python3 test/test_quantization.py TestQuantizeEagerOps.test_conv_transpose_op

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

terrychenism committed:
Update on "[Qunat] Add ConvTranspose reference module"


Summary:
Add ConvTranspose reference module

Test Plan:
python3 test/test_quantization.py TestQuantizeEagerOps.test_conv_transpose_op

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

terrychenism committed:
Update on "[Qunat] Add ConvTranspose reference module"


Summary:
Add ConvTranspose reference module

Test Plan:
python3 test/test_quantization.py TestQuantizeEagerOps.test_conv_transpose_op

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

terrychenism commented:
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34126542).

terrychenism committed:
Update on "[Qunat] Add ConvTranspose reference module"


Summary:
Add ConvTranspose reference module

Test Plan:
python3 test/test_quantization.py TestQuantizeEagerOps.test_conv_transpose_op

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34126542](https://our.internmc.facebook.com/intern/diff/D34126542)

[ghstack-poisoned]

terrychenism commented:
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34126542).

terrychenism committed:
Update on "[Qunat] Add ConvTranspose reference module"


Summary:
Add ConvTranspose reference module

Test Plan:
python3 test/test_quantization.py TestQuantizeEagerOps.test_conv_transpose_op

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34126542](https://our.internmc.facebook.com/intern/diff/D34126542)

[ghstack-poisoned]

terrychenism commented:
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34126542).

terrychenism committed:
Update on "[Qunat] Add ConvTranspose reference module"


Summary:
Add ConvTranspose reference module

Test Plan:
python3 test/test_quantization.py TestQuantizeEagerOps.test_conv_transpose_op

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34126542](https://our.internmc.facebook.com/intern/diff/D34126542)

[ghstack-poisoned]

terrychenism commented:
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34126542).

terrychenism committed:
Update on "[Qunat] Add ConvTranspose reference module"


Summary:
Add ConvTranspose reference module

Test Plan:
python3 test/test_quantization.py TestQuantizeEagerOps.test_conv_transpose_op

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34126542](https://our.internmc.facebook.com/intern/diff/D34126542)

[ghstack-poisoned]

terrychenism commented:
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34126542).

jerryzh168 commented on test/quantization/eager/test_quantize_eager_ptq.py:
```diff
@@ -74,6 +75,131 @@
+            nn.ConvTranspose2d: nnqr.ConvTranspose2d,
+            nn.ConvTranspose3d: nnqr.ConvTranspose3d,
+        }
+        ref_m = convert(ref_m, mapping=reference_module_mapping)
```
should we check the modules after the reference convert as well?

jerryzh168 reviewed (approved):
Looks good overall, had a few comments inline

terrychenism commented:
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34126542).

malfet commented:
Reverting as it breaks MacOS test sanity, see https://hud.pytorch.org/commit/pytorch/pytorch/7a031ec17f5cfc052e618694d4bee28381a6051d
```
test_conv_1d (quantization.eager.test_quantize_eager_ptq.TestQuantizeEagerOps) ... ERROR (0.016s)
```

terrychenism committed:
Update on "[Qunat] Add ConvTranspose reference module"


Summary:
Add ConvTranspose reference module

Test Plan:
python3 test/test_quantization.py TestQuantizeEagerOps.test_conv_transpose_op

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34126542](https://our.internmc.facebook.com/intern/diff/D34126542)

[ghstack-poisoned]

terrychenism commented:
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34292446).

terrychenism committed:
Update on "[Qunat] Add ConvTranspose reference module"


Summary:
Add ConvTranspose reference module

Test Plan:
python3 test/test_quantization.py TestQuantizeEagerOps.test_conv_transpose_op

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34292446](https://our.internmc.facebook.com/intern/diff/D34292446)

[ghstack-poisoned]
